stdlib/public/runtime/MetadataLookup.cpp

```diff
@@ -3318,44 +3347,106 @@ static void _gatherWrittenGenericArgs(
               return substitutions.getWitnessTable(type, index);
-            }).getType().getMetadata());
+            }).getType().getMetadata();
+      if (!genericArg)
+        return false;
+
+      genericArgs[*lhsFlatIndex] = MetadataOrPack(genericArg);
       continue;
     }
 
     // If we do have an argument for this parameter, it might be that
     // the right-hand side is itself a generic parameter, which means
     // we have a same-type constraint A == B where A is already filled in.
     auto rhsParam = demangleToGenericParamRef(req.getMangledTypeName());
-    if (!rhsParam)
+
+    // If the rhs parameter is not a generic parameter itself with
+    // (depth, index), it could potentially be some associated type. Check it
+    // again later once we've found all of the other same types.
+    if (!rhsParam) {
+      needToCheckAssociatedSameTypes = true;
       continue;
+    }
 
     auto rhsFlatIndex =
       _depthIndexToFlatIndex(rhsParam->first, rhsParam->second,
                              genericParamCounts);
-    if (!rhsFlatIndex || *rhsFlatIndex >= allGenericArgs.size())
+    if (!rhsFlatIndex || *rhsFlatIndex >= genericArgs.size())
+      return false;
+
+    if (genericArgs[*rhsFlatIndex] || !genericArgs[*lhsFlatIndex])
+      return false;
+
+    genericArgs[*rhsFlatIndex] = genericArgs[*lhsFlatIndex];
+  }
+
+  if (!needToCheckAssociatedSameTypes) {
+    return true;
+  }
+
+  // Once again, loop through our list and look for same type constraints where
+  // the rhs is an associated type of sorts.
+  for (const auto &req : genericContext->getGenericRequirements()) {
+    // We only care about same-type constraints.
+    if (req.Flags.getKind() != GenericRequirementKind::SameType) {
       continue;
+    }
+
+    auto lhsParam = demangleToGenericParamRef(req.getParam());
 
-    if (allGenericArgs[*rhsFlatIndex] || !allGenericArgs[*lhsFlatIndex])
+    if (!lhsParam) {
       continue;
+    }
+
+    auto lhsFlatIndex =
+      _depthIndexToFlatIndex(lhsParam->first, lhsParam->second,
+                             genericParamCounts);
+    if (!lhsFlatIndex || *lhsFlatIndex >= genericArgs.size())
+      return false;
+
+    auto rhsParam =
+        swift_getTypeByMangledName(MetadataState::Abstract,
+          req.getMangledTypeName(),
+          keyArgs.data(),
+          [&substitutions](unsigned depth, unsigned index) {
+            return substitutions.getMetadata(depth, index).Ptr;
+          },
+          [&substitutions](const Metadata *type, unsigned index) {
+            return substitutions.getWitnessTable(type, index);
+          }).getType().getMetadata();
+
+    if (!rhsParam) {
+      return false;
+    }
+
+    // If we already have an argument for the lhs, then just check that it is
+    // indeed == to the rhs type.
+    if (auto genericArg = genericArgs[*lhsFlatIndex]) {
+      if (genericArg.getMetadata() != rhsParam) {
+        return false;
```

Comment:
Aren't callers supposed to check the requirements with checkGenericRequirements()?